libavformat/httpauth.c

```diff
@@ -199,6 +221,143 @@ static char *make_digest_auth(HTTPAuthState *state, const char *username,
+    } else if (!strcmp(digest->qop, "auth-int")) {
+        /* qop=auth-int not supported */
+        return NULL;
+    } else {
+        /* Unsupported qop value. */
+        return NULL;
+    }
+
+    len = strlen(username) + strlen(state->realm) + strlen(digest->nonce) +
+              strlen(uri) + strlen(response) + strlen(digest->algorithm) +
+              strlen(digest->opaque) + strlen(digest->qop) + strlen(cnonce) +
+              strlen(nc) + 150;
+
+    authstr = av_malloc(len);
+    if (!authstr)
+        return NULL;
+    snprintf(authstr, len, "Authorization: Digest ");
+
+    /* TODO: Escape the quoted strings properly. */
+    av_strlcatf(authstr, len, "username=\"%s\"",   username);
+    av_strlcatf(authstr, len, ", realm=\"%s\"",     state->realm);
+    av_strlcatf(authstr, len, ", nonce=\"%s\"",     digest->nonce);
+    av_strlcatf(authstr, len, ", uri=\"%s\"",       uri);
+    av_strlcatf(authstr, len, ", response=\"%s\"",  response);
+
+    // we are violating the RFC and use "" because all others seem to do that too.
+    if (digest->algorithm[0])
+        av_strlcatf(authstr, len, ", algorithm=\"%s\"",  digest->algorithm);
+
+    if (digest->opaque[0])
+        av_strlcatf(authstr, len, ", opaque=\"%s\"", digest->opaque);
+    if (digest->qop[0]) {
+        av_strlcatf(authstr, len, ", qop=\"%s\"",    digest->qop);
+        av_strlcatf(authstr, len, ", cnonce=\"%s\"", cnonce);
+        av_strlcatf(authstr, len, ", nc=%s",         nc);
+    }
+
+    av_strlcatf(authstr, len, "\r\n");
+
+    return authstr;
+}
+
+/* Generate a digest reply, according to RFC 7616 and backcompatible to RFC 2617. */
+static char *make_sha_digest_auth(HTTPAuthState *state, const char *username,
+                              const char *password, const char *uri,
+                              const char *method)
+{
+    DigestParams *digest = &state->digest_params;
+    int len;
+    uint32_t cnonce_buf[4];
+    char cnonce[33];
+    char nc[9];
+    int i;
+    char A1hash[65], A2hash[65], response[65];
+    struct AVSHA *sha256ctx; //SHA-256
+    struct AVSHA512 *sha512ctx; //SHA-512-256
+    uint8_t hash[32];
+    char *authstr;
+
+    digest->nc++;
+    snprintf(nc, sizeof(nc), "%08x", digest->nc);
+
+    /* Generate a client nonce. */
+    for (i = 0; i < 4; i++)
+        cnonce_buf[i] = av_get_random_seed();
+    ff_data_to_hex(cnonce, (const uint8_t*) cnonce_buf, sizeof(cnonce_buf), 1);
+    cnonce[2*sizeof(cnonce_buf)] = 0;
+    if (!strcmp(digest->algorithm, "SHA-256") || !strcmp(digest->algorithm, "SHA-256-sess")) {
+        sha256ctx = av_sha_alloc();
+        if (!sha256ctx)
+            return NULL;
+        av_sha_init(sha256ctx, 256);
+        update_sha256_strings(sha256ctx, username, ":", state->realm, ":", password, NULL);
+        av_sha_final(sha256ctx, hash);
+        ff_data_to_hex(A1hash, hash, 32, 1);
+        A1hash[64] = 0;
+        if (!strcmp(digest->algorithm, "SHA-256-sess")) {
```

Comment:
This seems to be overwriting A1hash. Should this not be an else statement so we write A1hash only once?

Reply:
Not really. 
Initially:
A1Hash = username) ":" + realm) ":" + passwd  //must have for both SHA-256 and SHA-256-sess
If algorithm is "SHA-256-sess", 
A1hash = A1hash + ":" + nonce-prime) + ":" + cnonce //This is only needed for "SHA-256-sess"
So the implementation looks correct.